File: day13/Day13.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day13 {

    static long solve(List<String> input, boolean  part2){
        long totalCost = 0;
        Pattern numPattern = Pattern.compile("-?\\d+");
        for(int i = 0; i < input.size(); i += 3){
            String block = input.get(i) + input.get(i+1) + input.get(i+2);
            Matcher m = numPattern.matcher(block);
            long[] v = new long[6];
            int idx = 0;
            while(m.find()) v[idx++] = Long.parseLong(m.group());
            long Ax = v[0], Ay = v[1]; // X values
            long Bx = v[2], By = v[3]; // y values
            long Px = v[4], Py = v[5]; // prices

            if (part2){
                 Px += 10_000_000_000_000L;
                Py += 10_000_000_000_000L;
            }

            long D = Ax * By - Ay * Bx; // using crammers rule Determinant
            if (D==0)continue;

            long aNum = Px * By - Py * Bx; // y det
            long bNum = Ax * Py - Px * Ay ; // x det

            if(aNum % D != 0 || bNum % D != 0) continue;

            long a = aNum/D;
            long b = bNum/D;

            if(a < 0 || b < 0) continue;

            totalCost += 3 * a + b;
        }
        return totalCost;
    }

    public static void main(String[] args) throws Exception{
        List<String> input;
        try (BufferedReader br = new BufferedReader(new FileReader("day13_inputs.txt"))) {
            input = new ArrayList<>();
            String line;
            while((line = br.readLine()) != null){
                if(!line.isEmpty()) input.add(line);
            }
        }
        System.out.println("Part 1: " + solve(input, false));
        System.out.println("Part 2: " + solve(input, true));
    }
}

File: day13/day13.py
import re

def solve(lines, part2=False):
    total_cost = 0

    lines = [ line for line in lines if line.strip()]

    for i in range(0, len(lines), 3):
        block = lines[i] + lines[i+1] + lines[i+2]
        nums = list(map(int, re.findall(f"-?\\d+", block)))
        print(nums)
        Ax, Ay, Bx, By, Px, Py = nums

        if part2:
            Px += 10**13
            Py += 10**13
        D = Ax*By - Bx*Ay
        if D == 0:
            continue

        a_nums = Px*By - Py*Bx
        b_nums = Ax*Py - Ay*Px

        if a_nums % D != 0 or b_nums % D != 0:
            continue

        a = a_nums // D
        b = b_nums // D

        if a < 0 or b < 0:
            continue

        total_cost += 3*a + b
    return total_cost



if __name__ == "__main__":
    with open("day13_inputs.txt") as f:
        lines = f.readlines();
    
print("Part 1: ", solve(lines, False))
print("Part 2:", solve(lines, True))